### ISSUE TITLE
Problem of dependencies when rerunning jobs

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1132]
In GitLab by @ yruprich on Sep 19, 2023, 09:36

Hello @dbeltrankyl, @kinow, @eferre1,

### Autosubmit Version

I am using AS3.14.0

### Expid affected(If applicable)

For example a6gg

### Which task has issues? Where is the log(If applicable)

It is not specific to one task but rather to the workflow. Dependencies are not correctly accounted for. An example is the job `a6gg_19000101_fc0_14_SIM` that is launched at the same time as `a6gg_19000101_fc0_13_SIM` whereas `a6gg_19000101_fc0_14_SIM` should depend of `a6gg_19000101_fc0_13_SIM`.

### Summary

I set in my job.conf the following dependencies:
```
[SIM]
DEPENDENCIES = INI SIM-1 CLEAN-3 NCTIME-3
```
```
[CLEAN]
DEPENDENCIES = POST SIM SIM+1 CMOROCE CMORATM CMORATM+1 CMORLPJG CLEAN-1
```
However, while running the workflow, AS simplifies my dependencies and the SIM jobs end up being only dependent of CLEAN-3. This is not a big deal as long as I am running the workflow linearly. But this becomes an issue when I need to rerun a jobs in the middle of my workflow. Which is what happened when I wanted to rerun my SIM jobs 13, 14 and 15. Then those 3 SIM jobs were submitted at the same time (i.e. not respecting the dependencies I set up).

### Steps to reproduce

1. Run the workflow once
2. rerun jobs in the middle...

Thank you,

Yohan

**Comment by kinow** [comment]
In GitLab by @eferre1 on Sep 19, 2023, 10:48

Hi @ yruprich, I took a quick look at the graph and it looks like the dependencies are correctly set, since each SIM depends on the previous one. From the log `20230919_084635_setstatus.log` it looks like you set the jobs to `READY` instead of `WAITING`, which might have caused the issue that you had, since `READY` is the status of the jobs that can be submitted, and `WAITING` means that the job still has to wait for its dependencies completion.

I think this already happened in the past (Aude had a similar issue if I recall correctly), but I'm not sure if at the end it was decided to add a check or something to the `setstatus` command to not allow or to confirm that the jobs can be set to `READY`. If it was added maybe it is on the 3.15 version, but as said, I'm not sure. Can you confirm this @dbeltrankyl?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 10:54

I think @eferre1 is right, thanks for the debug

We were planning in 3.15, if I recall correctly, to put a flag to double-check that you really want as final status -t "READY" instead of -t "WAITING" so this can't happen.

 But is still not implemented

**Comment by kinow** [comment]
In GitLab by @ yruprich on Sep 19, 2023, 11:14

Hi both, thanks for your feedback.

I don't think this is the problem. I never use the `READY` status when I use `setstatus`, I always use `WAITING`, exactly for the reason you are mentioning. The log you are referring to @eferre1 was created by AS when I launched the workflow for the first time I believe. The logs related to the command `setstatus` are in the directory `/esarchive/autosubmit/a6gg/tmp/ASLOGS`. See for example: `/esarchive/autosubmit/a6gg/tmp/ASLOGS/20230918_171147_setstatus.log`, where you can see lines 60-62 that I set the SIM jobs 13, 14 and 15 to `WAITING`. The problem is coming from something else.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 11:42

Ah, okay, sorry, I'm afraid that I jumped to a conclusion with Eric comment

I see that in both versions, 3.14 and 3.15, the workflow shown in the create seems to have the correct dependencies.  

Does it only happen in running time? Because it would be strange. I have a simplified copy of your workflow with only 30 chunks ( to visualize it ) that should be enough to be triggered.

Edit: The thing is that the SIM-1 shouldn't  never be simplified so I don't know what exactly happen there

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 11:58

If it is happening only during the `autosubmit run`, or your dependencies in the autosubmit monitor are well-made? 

I see that in the GUI, the dependencies of the SIM are well-made. Each SIM depends on it's previous. So, apparently, they aren't being simplified. 

I've checked the workflow plot on both versions and also seems okay. 

So, it seems that only happens during autosubmit run but not sure how

**Comment by kinow** [comment]
In GitLab by @eferre1 on Sep 19, 2023, 12:06

So, in that case, during the `setstatus` that I mention something weird happened, and then the next `run` had the jobs as `READY`. From the messages, it looks like some interference with the retrials system. You can see below the logs were you changed one job to `WAITING` (the one to run first), but the other 2 being changed on that same command (which caused me the confusion about the setup of the jobs to `READY`):

- From `20230919_084635_setstatus.log`:
```
2023-09-19 08:46:52,572 Final status: WAITING
2023-09-19 08:46:52,573 List of jobs to change: a6gg_19000101_fc0_13_SIM
(···)
2023-09-19 08:46:53,116 Using existing valid structure.
2023-09-19 08:46:53,182 CHANGED: job: a6gg_19000101_fc0_13_SIM status to: WAITING
2023-09-19 08:46:53,372 Updating FAILED jobs
2023-09-19 08:46:53,372 Resetting job: a6gg_19000101_fc0_14_SIM status to: READY for retrial...
2023-09-19 08:46:53,372 Resetting job: a6gg_19000101_fc0_15_SIM status to: READY for retrial...
2023-09-19 08:46:53,373 Updating WAITING jobs
```
- From `20230919_084711_run.log`:
```
2023-09-19 08:49:06,678 Job a6gg_19000101_fc0_15_SIM is QUEUING
2023-09-19 08:49:06,679 Job a6gg_19000101_fc0_13_SIM is QUEUING
2023-09-19 08:49:06,680 Job a6gg_19000101_fc0_14_SIM is QUEUING
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 12:09

Yes, I just noticed that

I think there is a bug with the RETRIALS and setstatus in that version. 

If you put retrials to 0 for SIM jobs before performing setstatus it will work. Once the setstatus is completed, you can set the retrials again to > 0.

**Comment by kinow** [comment]
In GitLab by @ yruprich on Sep 19, 2023, 13:21

> put retrials to 0 for SIM jobs before performing setstatus it will work. Once the setstatus is completed, you can set the retrials again to > 0.

Thanks @dbeltrankyl, I will keep that in mind for the next times!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 14:38

Hello @ yruprich ,

It seems a different case, 

This one set status was done correctly.

However, In this one, I can see that the plot simplifies the sim-1 dependency. I'm afraid that in  this case, I can't think of a solution aside of put them in suspended status or cancel them manually, but I'll check what is happening to fix it in 3.15.0.

<details><summary>I'll save here the current jobs conf to don't forget</summary>

```
[LOCAL_SETUP]
FILE = templates/common/localsetup.tmpl.sh
PLATFORM = LOCAL
CHECK = true

[SYNCHRONIZE]
FILE = templates/common/synchronize.tmpl.sh
PLATFORM = transfer_node
DEPENDENCIES = LOCAL_SETUP
CHECK = true

[REMOTE_SETUP]
FILE = sources/runtime/autosubmit/compilation.sh
DEPENDENCIES = SYNCHRONIZE
WALLCLOCK = 2:00
PROCESSORS = 4
CUSTOM_DIRECTIVES = ["#SBATCH -p interactive"]
CHECK = true

[INI]
FILE = templates/common/ini.tmpl.sh
DEPENDENCIES = REMOTE_SETUP
RUNNING = member
WALLCLOCK = 00:30
CUSTOM_DIRECTIVES = ["#SBATCH -p interactive"]
CHECK = true
#QUEUE = debug

[SIM]
FILE = sources/runtime/autosubmit/ece-esm.sh
#DEPENDENCIES = INI SIM-1 CLEAN-26 NCTIME-26
DEPENDENCIES = INI SIM-1 CLEAN-3 NCTIME-3
RUNNING = chunk
WALLCLOCK = 2:00
PROCESSORS = 768
TASKS = 48
THREADS = 1
CHECK = on_submission
QUEUE = debug

[CMOROCE]
FILE = templates/common/cmoroce.tmpl.sh
DEPENDENCIES = SIM
WALLCLOCK = 0:15
RUNNING = chunk
TASKS = 8
CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]
CHECK = true
QUEUE = debug

[CMORATM]
FILE = templates/common/cmoratm.tmpl.sh
DEPENDENCIES = SIM SIM+1
WALLCLOCK = 0:15
RUNNING = chunk
CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]
CHECK = true
QUEUE = debug

[CMORLPJG]
FILE = templates/common/cmorlpjg.tmpl.sh
DEPENDENCIES = SIM
WALLCLOCK = 0:20
RUNNING = chunk
CUSTOM_DIRECTIVES =  ["#SBATCH --exclusive"]
CHECK = true
QUEUE = debug

[POST]
FILE = templates/common/post.tmpl.sh
DEPENDENCIES = CMORATM CMORATM+1 CMOROCE CMORLPJG SAVEIC SIM SIM+1 POST-1
RUNNING = chunk
WALLCLOCK = 00:30
# when using the classic outclass make sure you are running POST in exclusive mode (for MN4): CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]
CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]
CHECK = true
QUEUE = debug

[CLEAN]
FILE = templates/common/clean.tmpl.sh
PLATFORM = transfer_node
DEPENDENCIES = POST SIM SIM+1 CMOROCE CMORATM CMORATM+1 CMORLPJG CLEAN-1
RUNNING = chunk
WALLCLOCK = 00:30
CHECK = true

[EARTHDIAGS]
FILE = templates/common/earthdiags.tmpl.sh
PLATFORM = nord3v2
DEPENDENCIES = CLEAN
RUNNING = chunk
WALLCLOCK = 02:00
CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]
CHECK = true
NOTIFY_ON = FAILED

[SAVEIC]
FILE = templates/common/saveic.tmpl.sh
DEPENDENCIES = SIM
RUNNING = chunk
WALLCLOCK = 02:00
CHECK = true
THREADS = 6
QUEUE = debug

[CLEAN_MEMBER]
FILE = templates/common/clean_member.tmpl.sh
PLATFORM = transfer_node
DEPENDENCIES = CLEAN PERFORMANCE_METRICS SAVEIC
RUNNING = member
WALLCLOCK = 02:00
```

</details>

**Comment by kinow** [comment]
In GitLab by @ yruprich on Sep 19, 2023, 18:07

Maybe an important information since the problem I have for the expid `a6gg` and `a4wq` may not be the same.

For `a4wq`, I ran the workflow a first time (that had 121 chunks). Later on, I added 2 more SIM chunks to the workflow. Then the problem I have happens when I am trying to re-run the original last chunk (i.e. SIM 121) followed by the two newly added chunks (i.e. 122 and 123).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 20, 2024, 08:36

Hello, 

Finally, this won't be available in 3.15, as the changes are too core for a version that is in maintenance mode. 

For 4.1+, you can fine-tune the dependencies, and also, you can set a filter "all." It should be possible to add a redundant edge so the rerun is possible. Anyway, I'll open a new issue for 4.1 to test it ( https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1331)
